File: sorting/merge_sort.py
"""
Divides a given list into two halves, 
sorts those halves, and merges them in order.
"""

# Works better on linked lists


def merge_sort(lst):
    if len(lst) > 1:
        mid = len(lst) // 2
        left = lst[:mid]
        right = lst[mid:]

        merge_sort(left)
        merge_sort(right)

        merge(lst, left, right)


def merge(lst, left, right):
    i, j, k = 0, 0, 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            lst[k] = left[i]
            i += 1
        else:
            lst[k] = right[j]
            j += 1
        k += 1

    while i < len(left):
        lst[k] = left[i]
        i += 1
        k += 1

    while j < len(right):
        lst[k] = right[j]
        j += 1
        k += 1


if __name__ == '__main__':

    lst = [3, 2, 1, 5, 4]
    print("Unsorted list is:", lst)

    merge_sort(lst)
    print()
    print("Sorted list is: ", lst)

File: sorting/mergeSort.cpp
#include <vector>
#include <cmath>

/*
Divides a given array into two halves,
sorts those halves, and merges them in order.
*/

// Works better on linked lists


void mergeSort(std::vector<int>& arr)
{
    if (arr.size() > 1)
    {
        int mid = std::floor(arr.size() / 2);
        std::vector<int> left;
        std::vector<int> right;

        for (int i = 0; i < mid; i++)
        {
            left.push_back(arr[i]);
        }
        for (int i = mid; i < arr.size(); i++)
        {
            right.push_back(arr[i]);
        }

        mergeSort(left);
        mergeSort(right);

        merge(arr, left, right);
    }
}


void merge(std::vector<int>& arr, std::vector<int>& left, std::vector<int>& right)
{
    int i, j, k = 0;

    while (i < left.size() && j < right.size())
    {
        if (left[i] < right[j])
        {
            arr[k] = left[i];
            i++;
        }
        else
        {
            arr[k] = right[j];
            j++;
        }
        k++;
    }

    while (i < left.size())
    {
        arr[k] = left[i];
        i++;
        k++;
    }

    while (j < right.size())
    {
        arr[k] = right[j];
        j++;
        k++;
    }
}
